File: src - c++/Esercizi/Cifratore/cifratore.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

class Cifratore {
private:
  int chiave;
  char CifraCarattere(char c, bool cifra) const;
public:
  Cifratore(int c);
  void Cifra(char* str) const;
  void Decifra(char* str) const;
};

Cifratore::Cifratore(int c): chiave(c) {
}

char Cifratore::CifraCarattere(char c, bool cifra) const {
  if (cifra)
    return c + chiave;
  else
    return c - chiave;  
}

void Cifratore::Cifra(char* str) const{ 
  while (*str) {
    *str = CifraCarattere(*str, true);
    str++;
  }
}

void Cifratore::Decifra(char* str) const {
  while (*str) {
    *str = CifraCarattere(*str, false);
    str++;
  }  
}

int main()
{
  char str[100];
  int chiave;

  cout << "Inserisci la parola da cifrare: ";
  cin >> str;
  cout << "Inserisci la chiave di cifratura: ";
  cin >> chiave;
  
  Cifratore c(chiave);
    
  cout << "Stringa: " << str << endl; 
  c.Cifra(str);
  cout << "Cifratura: " << str << endl; 
  c.Decifra(str); 
  cout << "Decifratura: " << str << endl; 

  system("PAUSE");
  return 0;
}

File: src - java/Esercizi/Cifratore/Cifratore.java
public class Cifratore {
	public int chiave;

	public Cifratore(int _chiave) {
		chiave = _chiave;
	}

	private char cifraCarattere(char c, boolean cifra) {
		if (cifra)
			return (char) (c + chiave); // cifratura: aggiunge
		else
			return (char) (c - chiave); // decifratura: sottrae
	}

	public String cifra(String str) {
		// La classe String è immutabile. I suoi caratteri non
		// possono essere modificati. E' pertanto necessario
		// creare una nuova stringa contenente la sequenza dei
		// caratteri cifrati.
		// La classe StringBuilder del package java.lang
		// è utile a costruire stringhe in maniera incrementale.

		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < str.length(); i++)
			sb.append(cifraCarattere(str.charAt(i), true));
		return sb.toString();
	}

	public String decifra(String str) {
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < str.length(); i++)
			sb.append(cifraCarattere(str.charAt(i), false));
		return sb.toString();
	}
}